cmd/server.go

```diff
@@ -90,38 +91,78 @@ func bindServerArgs(cmd *cobra.Command, args []string) {
 	errz.Fatal(viper.BindPFlag("cpuprofile", cmd.Flags().Lookup("cpuprofile")))
 }
 
-// serverOptsFor returns server.Opts for the given command.
-func serverOptsFor(cmd *cobra.Command) (server.Opts, error) {
+// serverOptsFor returns map of plugins and server.Opts for the given command.
+func serverOptsFor(cmd *cobra.Command) (map[string]plugin.Root, server.Opts, error) {
 	// Read the config
 	configFile, err := cmd.Flags().GetString("config-file")
 	if err != nil {
 		panic(err.Error())
 	}
 	if err := config.ReadFrom(configFile); err != nil {
-		return server.Opts{}, err
+		return nil, server.Opts{}, err
 	}
 
 	// Unmarshal the external plugins, if any are specified
-	var externalPlugins []plugin.ExternalPluginSpec
+	var externalPlugins []externalPluginSpec
 	if err := viper.UnmarshalKey("external-plugins", &externalPlugins); err != nil {
-		return server.Opts{}, fmt.Errorf("failed to unmarshal the external-plugins key: %v", err)
+		return nil, server.Opts{}, fmt.Errorf("failed to unmarshal the external-plugins key: %v", err)
 	}
 
-	config := make(map[string]map[string]interface{})
-	for name := range internalPlugins {
-		config[name] = viper.GetStringMap(name)
+	// Copy internalPlugins so we don't mutate it.
+	plugins := make(map[string]plugin.Root)
+	for name, plug := range internalPlugins {
+		plugins[name] = plug
 	}
+
+	// Ensure external plugins are valid scripts and convert them to plugin.Root types.
 	for _, spec := range externalPlugins {
-		name := spec.Name()
+		intPlugin, err := spec.load()
+		if err != nil {
+			log.Warnf("%v failed to load: %+v", spec.Script, err)
+			continue
+		}
+
+		name := plugin.Name(intPlugin)
+		if _, ok := plugins[name]; ok {
+			log.Warnf("Overriding plugin %s with external plugin %s", name, spec.Script)
+		}
+		plugins[name] = intPlugin
+	}
+
+	config := make(map[string]map[string]interface{})
+	for name := range plugins {
 		config[name] = viper.GetStringMap(name)
 	}
 
 	// Return the options
-	return server.Opts{
-		CPUProfilePath:  viper.GetString("cpuprofile"),
-		LogFile:         viper.GetString("logfile"),
-		LogLevel:        viper.GetString("loglevel"),
-		PluginConfig:    config,
-		ExternalPlugins: externalPlugins,
+	return plugins, server.Opts{
+		CPUProfilePath: viper.GetString("cpuprofile"),
+		LogFile:        viper.GetString("logfile"),
+		LogLevel:       viper.GetString("loglevel"),
+		PluginConfig:   config,
 	}, nil
 }
+
+// externalPluginSpec represents an external plugin's specification.
+type externalPluginSpec struct {
```

Comment:
Hmm, I still think this should be in plugin. Otherwise if we add other options to this (like output format), that will be extra arguments to `NewExternalPluginRoot` that may eventually get translated to their own separate struct and thus result in redundant code.

Reply:
I think of externalPluginSpec is solely a serialization format. I don't think it has much to do with the actual plugin implementation.

I can kind of see your point though. NewExternalPluginRoot should probably take an ExternalPluginSpec, not a string (or two strings) as is currently implemented.

Reply:
I think of it as what clients need to know about external plugins in order to use them (it being a serialization format is just a nice consequence of that). That reason is why I liked the addition of `load` in the PR